core/src/program/blocks/span_block.rs

```diff
@@ -153,91 +145,146 @@ impl OpBatch {
     }
 }
 
-// HELPER FUNCTIONS
-// ================================================================================================
+/// An accumulator used in construction of operation batches.
+struct OpBatchAccumulator {
+    ops: Vec<Operation>,
+    groups: [Felt; BATCH_SIZE],
+    /// Value of the currently active op group.
+    group: u64,
+    /// Index of the next opcode in the current group.
+    op_idx: usize,
+    /// index of the current group in the batch.
+    group_idx: usize,
+    // Index of the next free group in the batch.
+    next_group_idx: usize,
+}
 
-fn batch_ops(ops: Vec<Operation>) -> (Vec<OpBatch>, Digest) {
-    // encodes a group of opcodes; up to 7 opcodes can fit into one group
-    let mut op_group = 0u64;
+impl OpBatchAccumulator {
+    /// Returns a blank [OpBatchAccumulator].
+    pub fn new() -> Self {
+        Self {
+            ops: Vec::new(),
+            groups: [Felt::ZERO; BATCH_SIZE],
+            group: 0,
+            op_idx: 0,
+            group_idx: 0,
+            next_group_idx: 1,
+        }
+    }
 
-    // index of the next opcode in the current group; valued range of values is [0, 8)
-    let mut op_idx = 0;
+    /// Returns true if this accumulator does not contain any operations.
+    pub fn is_empty(&self) -> bool {
+        self.ops.is_empty()
+    }
 
-    // indexes of the current group in the batch, and the next free group in the batch; these
-    // could be different by more than 1 if some operations in the batch carry immediate values.
-    let mut group_idx = 0;
-    let mut next_group_idx = 1;
+    /// Returns true if this accumulator can accept the specified operation.
+    pub fn can_accept_op(&self, op: Operation) -> bool {
+        // if the current group is full and the batch is full, return false
+        let next_group_idx = if self.op_idx == GROUP_SIZE {
+            if self.next_group_idx >= BATCH_SIZE {
+                return false;
+            }
+            self.next_group_idx + 1
+        } else {
+            self.next_group_idx
+        };
```

Comment:
I think in general everything is still a bit hard to follow, mostly because the underlying requirements of the span block are complicated, but  perhaps partly because of the way logic is split between `add_op` and `can_accept_op`. This makes sense in terms of the batch accumulation, but maybe adding a bit more documentation here about what the conditions are for accepting an op would be helpful. (I know it's at the top of the file, but a brief reminder alongside the code could help)